### PR TITLE
Refactor TLS code with a new `tlsconfig` package

### PR DESCRIPTION
**Pull request by tiborvass**
This patch creates a new `tlsconfig` package to handle creation of
secure-enough TLS configurations for clients and servers.

The package was created by refactoring TLS code in the client and the
daemon. After this patch, it is expected that all code creating TLS
configurations use this `tlsconfig` package for greater security,
consistency and readability.

On the server side, this fixes a bug where --tlsverify was not taken
into account. Now, if specified, it will require the client to
authenticate.

Signed-off-by: Tibor Vass tibor@docker.com
# 

Note: this patch does not modify code under `registry/` although it is expected for other patches touching code that creates TLS configurations (including `registry/`), to reuse this `tlsconfig` package.

Ping @diogomonica @NathanMcCauley @dmcgowan @docker/core-maintainers

### CODE REVIEW

**Comment by calavera** [comment]
do we really need a new package just for this configuration? maybe it should live within the `sockets` package. We could also rename that package to something else if it's not descriptive enough.

**Comment by tiborvass** [comment]
@calavera in my opinion, yes.

File: pkg/tlsutil/config.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	tlsConfig := NewConfig()
+	tlsCert, err := tls.LoadX509KeyPair(cert, key)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("Could not load X509 key pair (%s, %s): %v", cert, key, err)
+		}
+		return nil, fmt.Errorf("Error reading X509 key pair (%s, %s): %v. Make sure the key is encrypted.", cert, key, err)
+	}
+	tlsConfig.Certificates = []tls.Certificate{tlsCert}
+	if ca != "" {
```
Comment (tiborvass) [inline pkg/tlsutil/config.go]: I find this weird but it was like that: https://github.com/docker/docker/pull/13636/files#diff-d6f439c36bc4930ab43c13d5ad3868d2L58

I wonder if this shouldn't be `if ca != "" && verify {`

File: pkg/tlsutil/config.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		tlsConfig.RootCAs = certPool
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	if cert != "" && key != "" {
+		tlsCert, err := tls.LoadX509KeyPair(cert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Couldn't load X509 key pair: %v. Make sure the key is encrypted", err)
```
Comment (dmcgowan) [inline pkg/tlsutil/config.go]: Saw it was like this in the previous version, but pretty sure the key needs to be NOT encrypted

File: pkg/tlsutil/config.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		certPool, err := certPool(ca)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.RootCAs = certPool
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	if cert != "" && key != "" {
```
Comment (dmcgowan) [inline pkg/tlsutil/config.go]: This is a logic change, previously the cert and key would be stat'ed and any error would cause it to skip. I don't necessarily agree with that logic but it should be called out because now it will error if a file does not exist.

File: pkg/tlsutil/config.go (hunk 2 above)
Comment (diogomonica) [inline pkg/tlsutil/config.go]: Doesn't make sense it being encrypted.

Comment (thaJeztah) [inline pkg/tlsutil/config.go]: IIRC there was quite some confusion around this. The original text was something like;

".... Is the key encrypted?"

**Comment by diogomonica** [comment]
I don't see any issues with this. LGTM

File: pkg/tlsutil/config.go (hunk 1 above)
Comment (dmcgowan) [inline pkg/tlsutil/config.go]: The verify flag has no meaning in this context, if the CA is provided the client should always be verified against it. Using require any client certificate has no value unless you plan on using the certificate for something else after after the handshake.

File: pkg/tlsutil/config.go (hunk 2 above)
Comment (dmcgowan) [inline pkg/tlsutil/config.go]: @thaJeztah excellent recollection, https://github.com/docker/docker/commit/2ea6c2c264f0f43beb877a83aea3e6e6ad38ae5d

Comment (thaJeztah) [inline pkg/tlsutil/config.go]: Ah! 🎓

Now that all the right people are watching this; it would be good to finally settle this and use the right message (if it's wrong) :-)

**Comment by ehazlett** [comment]
Yes!!  Thanks!

File: pkg/tlsutil/config.go (hunk 2 above)
Comment (tiborvass) [inline pkg/tlsutil/config.go]: @thaJeztah @dmcgowan I assume there was a bad reformulation in https://github.com/docker/docker/pull/11212/files It probably should have been `Make sure your key is *not* encrypted`.

Comment (dmcgowan) [inline pkg/tlsutil/config.go]: Yes, `Make sure your key is *not* encrypted` is correct

File: docker/docker.go [hunk 4]
```diff
@@ -85,6 +83,12 @@ func main() {
 
 	setDefaultConfFlag(flTrustKey, defaultTrustKeyFile)
 
+	// Regardless of whether the user sets it to true or false, if they
+	// specify --tlsverify at all then we need to turn on tls
+	if *flTlsVerify {
+		*flTls = true
+	}
```
Comment (estesp) [inline docker/docker.go]: I think this is why you have a broken CI test on janky.. currently **TestRunTLSverify** assumes that specifying `--tls-verify` without enabling tls will cause a `docker run` failure, but it isn't anymore.

Comment (tiborvass) [inline docker/docker.go]: yes fixing now :)

File: pkg/tlsutil/config.go (hunk 2 above)
Comment (thaJeztah) [inline pkg/tlsutil/config.go]: Think *this* is where it actually went wrong; https://github.com/docker/docker/pull/11056#issuecomment-76467272 and https://github.com/docker/docker/pull/11073#discussion_r25570875. 😙

**Comment by nathanleclaire** [comment]
Nice.

File: pkg/tlsutil/config.go (hunk 3 above)
Comment (tiborvass) [inline pkg/tlsutil/config.go]: Here is how it looked like previously for the client: https://github.com/docker/docker/pull/13636/files#diff-b715a530d535012e418a564a3eafedb9L130

File: pkg/tlsutil/config.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("Couldn't load X509 key pair: %v. Make sure the key is not encrypted", err)
+		}
+		tlsConfig.Certificates = []tls.Certificate{tlsCert}
+	}
+	return tlsConfig, nil
+}
+
+// ServerConfig returns a TLS configuration meant to be used by a server.
+func ServerConfig(enable, verify bool, ca, cert, key string) (*tls.Config, error) {
```
Comment (tiborvass) [inline pkg/tlsutil/config.go]: @dmcgowan `verify` is never used here, should we remove it?

Comment (dmcgowan) [inline pkg/tlsutil/config.go]: I vote remove it, it is unused and without meaning

File: api/server/server.go [hunk 6]
```diff
@@ -44,11 +46,7 @@ type ServerConfig struct {
 	CorsHeaders string
 	Version     string
 	SocketGroup string
-	Tls         bool
-	TlsVerify   bool
-	TlsCa       string
-	TlsCert     string
-	TlsKey      string
+	TlsConfig   *tls.Config
```
Comment (stevvooe) [inline api/server/server.go]: `TLSConfig`

File: api/server/server.go [hunk 7]
```diff
@@ -1402,22 +1400,15 @@ func (s *Server) ping(version version.Version, w http.ResponseWriter, r *http.Re
 }
 
 func (s *Server) initTcpSocket(addr string) (l net.Listener, err error) {
```
Comment (stevvooe) [inline api/server/server.go]: `initTCPSocket`

File: pkg/sockets/tcp_socket.go [hunk 8]
```diff
@@ -2,68 +2,19 @@ package sockets
...
-	return &TlsConfig{
-		Verify:      verify,
-		Certificate: tlsCert,
-		Key:         tlsKey,
-		CA:          tlsCA,
-	}
-}
-
-func NewTcpSocket(addr string, config *TlsConfig, activate <-chan struct{}) (net.Listener, error) {
+func NewTcpSocket(addr string, tlsConfig *tls.Config, activate <-chan struct{}) (net.Listener, error) {
```
Comment (stevvooe) [inline pkg/sockets/tcp_socket.go]: `newTCPListener` (or mirror `net` package). Could we also add documentation here? What does the activate argument do?

**Comment by stevvooe** [comment]
@tiborvass This looks like a good change. I would prefer to call it `tlcconf` or something like that, rather than overload the "util" terminology.

**Comment by aluzzardi** [comment]
ping @vieux @abronan

This is awesome - we have exactly that same code in Swarm. We could replace it with this `pkg`.

I agree with @stevvooe - `tlsconfig` or something less generic.

**Comment by tiborvass** [comment]
@aluzzardi @stevvooe I can rename it to `tlsconfig` and have the functions be `New()`, `Client()` and `Server()`

File: pkg/sockets/tcp_socket.go (hunk 8 above)
Comment (tiborvass) [inline pkg/sockets/tcp_socket.go]: It's some weird logic for daemon startup. This is out of scope for this PR.

File: api/server/server.go (hunk 6 above)
Comment (tiborvass) [inline api/server/server.go]: Will fix

File: api/server/server.go (hunk 7 above)
Comment (tiborvass) [inline api/server/server.go]: out of scope

**Comment by tiborvass** [comment]
Renamed the package from `tlsutil` to `tlsconfig`, and changed the name of the functions as well to make them shorter.

**Comment by stevvooe** [comment]
There are still a few docs and naming issues but they don't matter.

LGTM!

File: pkg/tlsconfig/config.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		tlsConfig.RootCAs = certPool
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	if cert != "" && key != "" {
+		tlsCert, err := tls.LoadX509KeyPair(cert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Couldn't load X509 key pair: %v. Make sure the key is not encrypted", err)
```
Comment (thaJeztah) [inline pkg/tlsconfig/config.go]: OCD warning; change to "Could not load ...", to be consistent with the other errors? (Don't forget the related test)

Comment (tiborvass) [inline pkg/tlsconfig/config.go]: fixing

File: pkg/tlsconfig/config.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		tlsConfig.RootCAs = certPool
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	if cert != "" && key != "" {
+		tlsCert, err := tls.LoadX509KeyPair(cert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Could not load X509 key pair: %v. Make sure the key is not encrypted", err)
```
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: This is a logic change cf https://github.com/docker/docker/pull/13636#discussion_r31451932
(we used to continue on error)

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
its so much cleaner

File: pkg/tlsconfig/config.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+// NewConfig returns a secure-enough-by-default TLS configuration object.
```
Comment (LK4D4) [inline pkg/tlsconfig/config.go]: Wrong comment

File: pkg/tlsconfig/config.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Could not read CA certificate %s: %v", ca, err)
+	}
+	certPool.AppendCertsFromPEM(file)
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+func Client(enable, verify bool, ca, cert, key string) (*tls.Config, error) {
+	if !enable {
```
Comment (LK4D4) [inline pkg/tlsconfig/config.go]: I don't know. This looks bad, like you try to just split current code and not create good package. I'm not sure that such things should be in `pkg/`.

File: pkg/tlsconfig/config.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	file, err := ioutil.ReadFile(ca)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read CA certificate %s: %v", ca, err)
+	}
+	certPool.AppendCertsFromPEM(file)
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+func Client(enable, verify bool, ca, cert, key string) (*tls.Config, error) {
```
Comment (LK4D4) [inline pkg/tlsconfig/config.go]: Also probably too much positional arguments, you'll never remember right order, maybe makes sense to create config struct.

File: pkg/tlsconfig/config.go (hunk 10 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: @dmcgowan i feel like I found the only case where this is an actual breaking change.
It used to continue if the client certs/keys were not found (or had any kind of error as a matter of fact). But then would error out on the TLS handshake, either because the server has no TLS, or because the server requires client cert. So in those cases, it makes sense to error out early.

However, if the server was configured without CA and thus does not require a Client cert: https://github.com/docker/docker/pull/13636/files#diff-57f93555eddab72218d665bf6f7b239fR88, then failing to read the client certs used to be followed by a successful TLS handshake (without client cert verification), whereas now we would flat out error out.

File: pkg/tlsconfig/config.go (hunk 11 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: fixing

File: pkg/tlsconfig/config.go (hunk 13 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: i think it's fine, I agree it's really the maximum, but I'd rather keep it this way. Plus we already have tls.Config, so having tlsconfig.Config is even more confusing.

File: pkg/tlsconfig/config.go (hunk 12 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: Maintainers of other repos already expressed their enthusiasm to having this a package under `pkg`. I agree it can be improved to be even less docker-y, but I'm sure we can do that in other PRs as the needs rise.

Comment (stevvooe) [inline pkg/tlsconfig/config.go]: @LK4D4 @tiborvass I agree with both of you. `bool` as function arguments are a broken pattern. The meaning of `enable` and `verify` have a space that might not be valid for all values of a TLS configuration. The meaning of the strings `ca, `cert` and `key` are also unclear (paths? raw data? what format?). This is not a good, generalized interface and needs work. We may want an option-style, _al a carte_ function signature, or something else.

On the other, it is arguable whether that is the aim of the PR. Much of the above criticism adds to the complexity but doesn't attack the goal. At least this interface demonstrates the problem and it might be problematic to create a better interface to communicate `verify` and `enable`. At least, we now have this in a single spot.

Comment (tiborvass) [inline pkg/tlsconfig/config.go]: Couldn't agree more, I still suggest we get this in, and feel free to change the interfaces in subsequent PRs as the need comes. Like @stevvooe said, I want one single place where we do everything TLS in a way that is easily reviewable. If the problem is that it's still too dockery and does not deserve being in `pkg`, (although there are counterexamples), I don't mind putting under `docker/tlsconfig` for now until we figure out an interface that works for everyone. /cc @ehazlett @aluzzardi

**Comment by tiborvass** [comment]
People need to be aware of https://github.com/docker/docker/pull/13636#discussion_r31476324

**Comment by estesp** [comment]
So, I think I may have confused myself between the various modes claimed to be supported in the docs: http://docs.docker.com/articles/https/#other-modes

So, given I can't quite be sure the breaking change doesn't disallow one of those modes (like your comment @tiborvass that says now even if the server isn't set up to do client auth it will error out early trying to read client cert), I'll let you tell me if I'm missing the impact of the change :)  If I am right that one of those modes is blocked from operating now, we need to fix docs and decide that is the right strategy.

If I'm totally off base, then just ignore me :)

**Comment by tiborvass** [comment]
@estesp thanks for the docs pointer, will look into it.

File: pkg/tlsconfig/config.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+			return nil, fmt.Errorf("Could not load X509 key pair: %v. Make sure the key is not encrypted", err)
+		}
+		tlsConfig.Certificates = []tls.Certificate{tlsCert}
+	}
+	return tlsConfig, nil
+}
+
+// Server returns a TLS configuration meant to be used by a server.
+func Server(enable bool, ca, cert, key string) (*tls.Config, error) {
+	if !enable {
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: What is the point of enable?

File: pkg/tlsconfig/config.go [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+		},
+	}
+}
+
+// certPool returns an X.509 certificate pool from `ca`, the certificate file.
+func certPool(ca string) (*x509.CertPool, error) {
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: s/ca/caFile/

Ideally, change to a typesafe param instead of `string`

File: pkg/tlsconfig/config.go (hunk 14 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: if false it returns no TLS config. It's basically `--tls` in docker.

File: pkg/tlsconfig/config.go [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+	// If we should verify the server, we need to load a trusted ca
+	certPool := x509.NewCertPool()
+	file, err := ioutil.ReadFile(ca)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read CA certificate %s: %v", ca, err)
+	}
+	certPool.AppendCertsFromPEM(file)
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: needs error check

Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: Also, please log the certs that are being trusted at a debug level.

File: pkg/tlsconfig/config.go [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+func Client(enable, verify bool, ca, cert, key string) (*tls.Config, error) {
+	if !enable {
+		return nil, nil
+	}
+	tlsConfig := New()
+	if verify {
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: I want to see a default version of this function that *always* does verification. Then, there can be another version of it is that is called `InsecureSSLClient` that skips verification.

File: pkg/tlsconfig/config.go (hunk 13 above)
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: The general problem of TLS configuration begs for a config struct or a builder pattern.

File: pkg/tlsconfig/config.go (hunk 14 above)
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: Ok, shouldn't that be dealt with outside of this pkg?

Comment (aluzzardi) [inline pkg/tlsconfig/config.go]: Agreed, this is preferable:

```
if enable {
  tlsconfig.Server(...)
}

File: pkg/tlsconfig/config.go (hunk 13 above)
Comment (stevvooe) [inline pkg/tlsconfig/config.go]: The right way to go here, for Go, is probably a "prototype" struct:

```
var DefaultTLSConfig = tls.Config{
		// Avoid fallback to SSL protocols < TLS1.0
		MinVersion:               tls.VersionTLS10,
		PreferServerCipherSuites: true,
		CipherSuites: []uint16{
			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
		},
}
```

Then, one can just copy during construction:

```
c := &TLConfig{}
*c = DefaultTLSConfig
```

One does need to be careful that certain fields aren't modified, such as `CipherSuites`. If so, they need to be copied.

However, as I said elsewhere on this PR, the right pattern here is to have options functions:

```
type TLSOption func(*TLSConfig) error

func BuildTLSConfig(options... TLSOption) (*TLSConfig, error) { ... }

var Insecure = TLSOption(func(c *TLSConfig) error {
    c.InsecureSkipVerify = true
})

func LoadCA(p string) TLSOption {
    return TLSOption(func(c *TLSConfig) error {
        // load file and then set CA on config
    })
}

func LoadKey(p string) TLSOption { ... }

// call it like this
c := BuildTLSConfig(Insecure, LoadCA("/etc/ca"), LoadKey("/etc/key.pem"))
```

This level of change might be outside the scope of this PR.

**Comment by tiborvass** [comment]
@NathanMcCauley @LK4D4 @stevvooe updated

File: docker/flags.go [hunk 18]
```diff
@@ -85,19 +86,16 @@ func getDaemonConfDir() string {
 }
 
 var (
-	flVersion   = flag.Bool([]string{"v", "-version"}, false, "Print version information and quit")
-	flDaemon    = flag.Bool([]string{"d", "-daemon"}, false, "Enable daemon mode")
-	flDebug     = flag.Bool([]string{"D", "-debug"}, false, "Enable debug mode")
-	flLogLevel  = flag.String([]string{"l", "-log-level"}, "info", "Set the logging level")
-	flTls       = flag.Bool([]string{"-tls"}, false, "Use TLS; implied by --tlsverify")
-	flHelp      = flag.Bool([]string{"h", "-help"}, false, "Print usage")
-	flTlsVerify = flag.Bool([]string{"-tlsverify"}, dockerTlsVerify, "Use TLS and verify the remote")
```
Comment (tiborvass) [inline docker/flags.go]: i only removed tlsverify here, the rest is spacing difference

File: pkg/tlsconfig/config.go [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+	tlsConfig.InsecureSkipVerify = options.SkipVerify
+	if !options.SkipVerify {
+		CAs, err := certPool(options.CAFile)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.RootCAs = CAs
+	}
+
+	if options.CertFile != "" && options.KeyFile != "" {
```
Comment (dmcgowan) [inline pkg/tlsconfig/config.go]: This check is insufficient since these values get defaulted to ~/.docker/cert.pem and ~/.docker/key.pem,

File: pkg/tlsconfig/config.go [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+	tlsConfig := DefaultTLSConfig
+	tlsCert, err := tls.LoadX509KeyPair(options.CertFile, options.KeyFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("Could not load X509 key pair (%s, %s): %v", options.CertFile, options.KeyFile, err)
+		}
+		return nil, fmt.Errorf("Error reading X509 key pair (%s, %s): %v. Make sure the key is not encrypted.", options.CertFile, options.KeyFile, err)
+	}
+	tlsConfig.Certificates = []tls.Certificate{tlsCert}
+	if options.CAFile != "" {
```
Comment (dmcgowan) [inline pkg/tlsconfig/config.go]: Same as above, this value gets defaulted to ~/.docker/ca.pem

File: pkg/tlsconfig/config.go (hunk 19 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: right, that's what i thought was the problem. Will find a fix.

File: pkg/tlsconfig/config.go [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	logrus.Debugf("Certificates: %v", pem)
+	if !certPool.AppendCertsFromPEM(pem) {
+		return nil, fmt.Errorf("failed to append certificates from PEM file: %s", caFile)
+	}
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+func Client(options Options) (*tls.Config, error) {
```
Comment (stevvooe) [inline pkg/tlsconfig/config.go]: This isn't quite the interface I had in mind. Ideally, if making this change, it would have functional options.

Comment (tiborvass) [inline pkg/tlsconfig/config.go]: @stevvooe I know, but this should be good enough, don't you think?

**Comment by tiborvass** [comment]
@stevvooe @dmcgowan @estesp I fixed the logic around client authentication for both client (bug in this PR but not in master), and on the server (bug also on master)

**Comment by dmcgowan** [comment]
LGTM

File: pkg/tlsconfig/config.go [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+// Options represents the information needed to create client and server TLS configurations.
+type Options struct {
+	SkipVerify bool
+	CAFile     string
+	CertFile   string
+	KeyFile    string
+}
+
+// New returns a secure-enough-by-default TLS configuration object.
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: Comment seems slightly out of place.

Comment (tiborvass) [inline pkg/tlsconfig/config.go]: thank you will fix

**Comment by tiborvass** [comment]
Fixed description comment of `DefaultTLSConfig` and also renamed it to `Default` since it is already in a package named `tlsconfig`.

File: pkg/tlsconfig/config.go [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+
+// certPool returns an X.509 certificate pool from `ca`, the certificate file.
+func certPool(caFile string) (*x509.CertPool, error) {
+	// If we should verify the server, we need to load a trusted ca
+	certPool := x509.NewCertPool()
+	pem, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read CA certificate %s: %v", caFile, err)
+	}
+	logrus.Debugf("Certificates: %v", pem)
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: Would make more sense to log the parsed certs, not the raw pem file.

File: pkg/tlsconfig/config.go [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+		return nil, fmt.Errorf("failed to append certificates from PEM file: %s", caFile)
+	}
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+// If enforceClientAuth is true and an error is encountered when loading the x509 certificate-key pair,
+// the error will be returned. Otherwise, it is ignored and the client will not authenticate.
+func Client(options Options, enforceClientAuth bool) (*tls.Config, error) {
+	tlsConfig := DefaultTLSConfig
```
Comment (NathanMcCauley) [inline pkg/tlsconfig/config.go]: yay for secure by default!

File: pkg/tlsconfig/config.go (hunk 23 above)
Comment (dmcgowan) [inline pkg/tlsconfig/config.go]: Ahh yeah, missed this in the logs, prints out a large hex array

**Comment by tiborvass** [comment]
Squashed commits and fixed the debug logging in the tlsconfig package. I also split SkipVerify to `InsecureSkipVerify` (for clients) and `ClientAuth` (for servers).

File: pkg/tlsconfig/config.go (hunk 19 above)
Comment (tiborvass) [inline pkg/tlsconfig/config.go]: @dmcgowan I put it back, but at the same time I fixed it outside of the package on @NathanMcCauley's request: https://github.com/docker/docker/pull/13636/files#diff-b715a530d535012e418a564a3eafedb9R113

**Comment by NathanMcCauley** [comment]
LGTM

File: docker/docker.go [hunk 25]
```diff
@@ -94,59 +99,35 @@ func main() {
...
-	if *flTlsVerify {
-		certPool := x509.NewCertPool()
-		file, err := ioutil.ReadFile(*flCa)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Couldn't read ca cert %s: %s\n", *flCa, err)
-			os.Exit(1)
+	var tlsConfig *tls.Config
+	if *flTls {
+		tlsOptions.InsecureSkipVerify = !*flTlsVerify
+		if flag.IsSet("-tlscert") {
```
Comment (dmcgowan) [inline docker/docker.go]: This logic is backwards, it should only check this if the flag is NOT set. When the flag is set, then not exists  should cause an error later on. This should only check if the default file exists if the flag is not set.

Comment (tiborvass) [inline docker/docker.go]: nice catch... fixing

**Comment by icecrime** [comment]
@jfrazelle Still LGTM for you?

**Comment by jfrazelle** [comment]
re LGTM

**Comment by icecrime** [comment]
Waiting on CI.

**Comment by estesp** [comment]
I'm testing the actual options based on the docs list of supported settings. But if someone else has already finished that then I'm fine with the existing LGTMs

**Comment by icecrime** [comment]
Well, tests fail: back to `2-needs-code-review`, ping @tiborvass.

**Comment by tiborvass** [comment]
Fixed the failing test.

**Comment by dmcgowan** [comment]
LGTM

**Comment by ewindisch** [comment]
Great. LGTM, although I'd love to see the server and client have separate defaults. That could easily come in a follow-up patch.
